Test/civl/inductive-sequentialization/ProducerConsumer.bpl

```diff
@@ -1,17 +1,23 @@
 // RUN: %parallel-boogie "%s" > "%t"
 // RUN: %diff "%s.expect" "%t"
 
-type {:linear "pid"} Pid = int;
-var {:layer 0,3} C:[int]int; // FIFO channel
-var {:layer 0,3} head:int;   // head index of the channel
-var {:layer 0,3} tail:int;   // tail index of the channel
+type {:datatype} Channel; // FIFO channel
+function {:constructor} Channel(C: [int]int, head: int, tail: int): Channel;
 
-const unique prod_id:int;
-const unique cons_id:int;
+type ChannelId; // id of a FIFO channel
+
+type {:linear "cid"} {:datatype} ChannelHandle; // permission for sending to or receiving from a channel
+function {:constructor} Send(cid: ChannelId): ChannelHandle;
+function {:constructor} Receive(cid: ChannelId): ChannelHandle;
+function {:inline} {:linear "cid"} ChannelIdCollector(cid: ChannelId) : [ChannelHandle]bool {
+  MapConst(false)[Send(cid) := true][Receive(cid) := true]
+}
+
+var {:layer 0,3} channels: [ChannelId]Channel; // pool of FIFO channels
 
 type {:pending_async}{:datatype} PA;
-function {:constructor} PRODUCER(x:int, pid:int) : PA;
-function {:constructor} CONSUMER(x:int, pid:int) : PA;
+function {:constructor} PRODUCER(x: int, cid: ChannelId, send_handle: ChannelHandle) : PA;
+function {:constructor} CONSUMER(x: int, cid: ChannelId, receive_handle: ChannelHandle) : PA;
```

Comment:
Would it simplify things if we'd eliminate the `cid` parameter, and instead extract the channel id from `ChannelHandle`?

Reply:
Done.